cshatag.1
.\"Generate README file for github: MANWIDTH=80 man ./cshatag.1 > README
.TH CSHATAG 1 "MAY 2012" Linux "User Manuals"
.SH NAME

cshatag \- compiled shatag

.SH SYNOPSIS

cshatag [OPTIONS] FILE [FILE...]

.SH DESCRIPTION

cshatag is a minimal and fast re-implementation of shatag
.br
( https://github.com/maugier/shatag , written in Python
by Maxime Augier )
.br
in a compiled language (since v2.0: Go, earlier versions: C).

cshatag is a tool to detect silent data corruption. It writes
the mtime and the sha256 checksum of a file into the file's
extended attributes. The filesystem needs to be mounted with
user_xattr enabled for this to work.
When run again, it compares stored mtime and checksum. If it
finds that the mtime is unchanged but the checksum has changed,
it warns on stderr and reports a <corrupt> status.
In any case, the status of the file is printed to stdout.

File statuses that appear on stdout are:
.br
	<new>         file is missing both attributes
.br
	<outdated>    both mtime and checksum have changed
.br
	<ok>          both checksum and mtime stayed the same
.br
	<timechange>  only mtime has changed, checksum stayed the same
.br
	<corrupt>     mtime stayed the same but checksum changed

When a file is considered <corrupt>, the stored checksum is
only updated if "-fix" is passed. For all other statuses
except <ok>, the stored checksum is always updated.

cshatag aims to be format-compatible with shatag and uses the
same extended attributes (see the COMPATIBILITY section).

cshatag was written in C in 2012 and has been
rewritten in Go in 2019.

.SH OPTIONS

-cpuprofile save cpu profile to specified file. This is useful
            for diagnosing performance issues.
.br
-dry-run    don't make any changes
.br
-fix        fix the stored sha256 on corrupt files
.br
-j          number of threads. Values <= 0 means auto
            (https://pkg.go.dev/github.com/charlievieth/fastwalk#DefaultNumWorkers).
.br
-q          quiet mode - don't report <ok> files
.br
-qq         quiet2 mode - only report <corrupt> files and errors
.br
-recursive  recursively process the contents of directories
.br
-remove     remove cshatag's xattrs from FILE


.SH EXAMPLES

Check all regular files in the file tree below the current working directory:
.br
# cshatag -qq -recursive .
.br
Errors like corrupt files will be printed to stderr.
Run without "-qq" to see progress output.

To remove the extended attributes from all files:
.br
# cshatag -recursive -remove .

.SH "RETURN VALUE"

0 Success
.br
1 Wrong number of arguments
.br
2 One or more files could not be opened
.br
3 One or more files is not a regular file
.br
4 Extended attributes could not be written to one or more files
.br
5 At least one file was found to be corrupt
.br
6 More than one type of error occurred

.SH COMPATIBILITY

cshatag writes the user.shatag.ts field with full integer
nanosecond precision, while python uses a double for the
whole mtime and loses the last few digits.

.SH AUTHOR
Jakob Unterwurzacher <jakobunt@gmail.com>
.br
https://github.com/rfjakob/cshatag

.SH COPYRIGHT
Copyright 2012 Jakob Unterwurzacher. MIT License.

.SH "SEE ALSO"
shatag(1), sha256sum(1), getfattr(1), setfattr(1)


CHANGELOG.md
Changelog
---------

*Short changelog - for all the details look at the git log.*

v2.2.1, 2024-08-23
* Fix `Makefile` to ensure the correct version string is baked
  into the binary ([#29](https://github.com/rfjakob/cshatag/issues/29))

v2.2.0, 2024-08-23
* Do not update corrupt files unless `-fix` is passed ([#28](https://github.com/rfjakob/cshatag/pull/28),
  **behavoir change**)
* Linux: use 100ns resolution when comparing timestamps instead of 1ns
  to match SMB protocol restrictions
  ([#21](https://github.com/rfjakob/cshatag/issues/21)
  [commit](https://github.com/rfjakob/cshatag/commit/3e1f62b38b493b2be75437c208ae7b1d6a90c8e8))
* MacOS: use 1s resolution when comparing timestamps to match
  MacOS SMB client restrictions ([#21](https://github.com/rfjakob/cshatag/issues/21))

v2.1.0, 2022-10-22
* Add `-dry-run` [#22](https://github.com/rfjakob/cshatag/issues/22)
* This version is called `v2.1.0` as opposed to `v2.1` to conform
  to go.mod versioning rules (three-digit semver).

v2.0, 2020-11-15
* Rewrite cshatag in Go
* add MacOS support
* Add `-remove` flag
* Add `-q` and `-qq` flags
* Accept multiple files per invocation to improve performance
* Work around problems on MacOS SMB mounts
  ([#11](https://github.com/rfjakob/cshatag/pull/11))

v1.1, 2019-06-09
* Add test suite (`make test`)
  ([commit](https://github.com/rfjakob/cshatag/commit/74496854e5c934b6809e816b9e854c5c6585a0f4))
* Add Travis CI
* Drop useless trailing null byte from `user.shatag.sha256`

v1.0, 2019-01-02
* Add `make format` target

2019-02-01
* Fix missing null termination in ts buffer that could lead
  to false positives
  ([commit](https://github.com/rfjakob/cshatag/commit/26873dd71656730d5744efb7fa595d529b3c9ae6))

2017-05-04
* Respect `PREFIX` for `make install`
  ([commit](https://github.com/rfjakob/cshatag/commit/8d1225aabb7bdd3750f161133931b1c456bc2fdb))

2016-09-17
* Check for malloc returning NULL
  ([commit](https://github.com/rfjakob/cshatag/commit/ecadbddffb5e23811a9ae4a5265c287d5ae5c151))

2012-12-05
* C source code & man page published on Github
  ([commit](https://github.com/rfjakob/cshatag/commit/5ce7674ea3210fd0bb6b06a81ca8823e0664761a))


cshatag.go
package main

import (
	"bytes"
	"crypto/sha256"
	"errors"
	"flag"
	"fmt"
	"io"
	"os"
	"runtime"
	"strconv"
	"strings"
	"syscall"

	"github.com/pkg/xattr"
)

const xattrSha256 = "user.shatag.sha256"
const xattrTs = "user.shatag.ts"
const zeroSha256 = "0000000000000000000000000000000000000000000000000000000000000000"

var GitVersion = ""
var removeFlag bool

type fileTimestamp struct {
	s  uint64
	ns uint32
}

func (ts *fileTimestamp) prettyPrint() string {
	return fmt.Sprintf("%010d.%09d", ts.s, ts.ns)
}

type fileAttr struct {
	ts     fileTimestamp
	sha256 []byte
}

func (a *fileAttr) prettyPrint() string {
	return fmt.Sprintf("%s %s", string(a.sha256), a.ts.prettyPrint())
}

// getStoredAttr reads the stored extendend attributes from a file. The file
// should look like this:
//
//     $ getfattr -d foo.txt
//     user.shatag.sha256="dc9fe2260fd6748b29532be0ca2750a50f9eca82046b15497f127eba6dda90e8"
//     user.shatag.ts="1560177334.020775051"
func getStoredAttr(f *os.File) (attr fileAttr, err error) {
	attr.sha256 = []byte(zeroSha256)
	val, err := xattr.FGet(f, xattrSha256)
	if err == nil {
		copy(attr.sha256, val)
	}
	val, err = xattr.FGet(f, xattrTs)
	if err == nil {
		parts := strings.SplitN(string(val), ".", 2)
		attr.ts.s, _ = strconv.ParseUint(parts[0], 10, 64)
		if len(parts) > 1 {
			ns64, _ := strconv.ParseUint(parts[1], 10, 32)
			attr.ts.ns = uint32(ns64)
		}
	}
	return attr, nil
}

// getMtime reads the actual modification time of file "f" from disk.
func getMtime(f *os.File) (ts fileTimestamp) {
	fi, err := f.Stat()
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
	if !fi.Mode().IsRegular() {
		fmt.Printf("Error: %q is not a regular file\n", f.Name())
		os.Exit(3)
	}
	ts.s = uint64(fi.ModTime().Unix())
	ts.ns = uint32(fi.ModTime().Nanosecond())
	return
}

// getActualAttr reads the actual modification time and hashes the file content.
func getActualAttr(f *os.File) (attr fileAttr, err error) {
	attr.sha256 = []byte(zeroSha256)
	attr.ts = getMtime(f)
	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		fmt.Println(err)
		os.Exit(2)
	}
	// Check if the file was modified while we were computing the hash
	ts2 := getMtime(f)
	if attr.ts != ts2 {
		return attr, syscall.EINPROGRESS
	}
	attr.sha256 = []byte(fmt.Sprintf("%x", h.Sum(nil)))
	return attr, nil
}

// storeAttr stores "attr" into extended attributes.
// Should look like this afterwards:
//
//     $ getfattr -d foo.txt
//     user.shatag.sha256="dc9fe2260fd6748b29532be0ca2750a50f9eca82046b15497f127eba6dda90e8"
//     user.shatag.ts="1560177334.020775051"
func storeAttr(f *os.File, attr fileAttr) (err error) {
	if runtime.GOOS == "darwin" {
		// SMB or MacOS bug: when working on an SMB mounted filesystem on a Mac, it seems the call
		// to `fsetxattr` does not update the xattr but removes it instead. So it takes two runs
		// of `cshatag` to update the attribute.
		// To work around this issue, we remove the xattr explicitly before setting it again.
		// https://github.com/rfjakob/cshatag/issues/8
		removeAttr(f)
	}
	err = xattr.FSet(f, xattrTs, []byte(attr.ts.prettyPrint()))
	if err != nil {
		return
	}
	err = xattr.FSet(f, xattrSha256, attr.sha256)
	return
}

// removeAttr removes any previously stored extended attributes. Returns an error
// if removal of either the timestamp or checksum xattrs fails.
func removeAttr(f *os.File) error {
	err1 := xattr.FRemove(f, xattrTs)
	err2 := xattr.FRemove(f, xattrSha256)
	if err1 != nil && err2 != nil {
		return errors.New(err1.Error() + "  " + err2.Error())
	}
	if err1 != nil {
		return err1
	}
	if err2 != nil {
		return err2
	}
	return nil
}

// printComparison prints something like this:
//
//     stored: faa28bfa6332264571f28b4131b0673f0d55a31a2ccf5c873c435c235647bf76 1560177189.769244818
//     actual: dc9fe2260fd6748b29532be0ca2750a50f9eca82046b15497f127eba6dda90e8 1560177334.020775051
func printComparison(stored fileAttr, actual fileAttr) {
	fmt.Printf(" stored: %s\n actual: %s\n", stored.prettyPrint(), actual.prettyPrint())
}

var stats struct {
	total      int
	errors     int
	inprogress int
	corrupt    int
	timechange int
	outdated   int
	ok         int
}

func checkFile(fn string) {
	stats.total++
	f, err := os.Open(fn)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %s\n", err)
		stats.errors++
		return
	}
	defer f.Close()

	if removeFlag {
		if err = removeAttr(f); err != nil {
			fmt.Fprintf(os.Stderr, "Error: %s\n", err)
			stats.errors++
			return
		}
		fmt.Printf("<removed xattr> %s\n", fn)
		stats.ok++
		return
	}

	stored, _ := getStoredAttr(f)
	actual, err := getActualAttr(f)
	if err == syscall.EINPROGRESS {
		fmt.Printf("<concurrent modification> %s\n", fn)
		stats.inprogress++
		return
	}
	if stored.ts == actual.ts {
		if bytes.Equal(stored.sha256, actual.sha256) {
			fmt.Printf("<ok> %s\n", fn)
			stats.ok++
			return
		}
		fmt.Fprintf(os.Stderr, "Error: corrupt file %q\n", fn)
		fmt.Printf("<corrupt> %s\n", fn)
		stats.corrupt++
	} else if bytes.Equal(stored.sha256, actual.sha256) {
		fmt.Printf("<timechange> %s\n", fn)
		stats.timechange++
	} else {
		// timestamp is outdated
		fmt.Printf("<outdated> %s\n", fn)
		stats.outdated++
	}
	printComparison(stored, actual)
	err = storeAttr(f, actual)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %s\n", err)
		stats.errors++
		return
	}
}

func main() {
	const myname = "cshatag"

	if GitVersion == "" {
		GitVersion = "(version unknown)"
	}

	flag.BoolVar(&removeFlag, "remove", false, "Remove any previously stored extended attributes.")
	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "%s %s\n", myname, GitVersion)
		fmt.Fprintf(os.Stderr, "Usage: %s [OPTION] FILE [FILE ...]\n", myname)
		fmt.Fprintf(os.Stderr, "Options:\n")
		flag.PrintDefaults()
		os.Exit(1)
	}
	flag.Parse()

	if flag.NArg() == 0 {
		flag.Usage()
	}

	for _, fn := range flag.Args() {
		checkFile(fn)
	}
	if (stats.ok + stats.outdated + stats.timechange) == stats.total {
		os.Exit(0)
	}
	if stats.corrupt > 0 {
		os.Exit(5)
	}
	os.Exit(2)
}
